go/x509util/x509util.go

```diff
@@ -0,0 +1,576 @@
+	OidExtensionSubjectDirectoryAttributes = asn1.ObjectIdentifier{2, 5, 29, 9}
+	OidExtensionInhibitAnyPolicy           = asn1.ObjectIdentifier{2, 5, 29, 54}
+	OidExtensionPolicyConstraints          = asn1.ObjectIdentifier{2, 5, 29, 36}
+	OidExtensionPolicyMappings             = asn1.ObjectIdentifier{2, 5, 29, 33}
+	OidExtensionFreshestCRL                = asn1.ObjectIdentifier{2, 5, 29, 46}
+
+	OidExtensionAuthorityInfoAccess = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 1}
+	OidExtensionSubjectInfoAccess   = asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 11}
+
+	// RFC 6962 s3.1
+	OidExtensionCTPoison = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3}
+	// RFC 6962 s3.3
+	OidExtensionCTSCT = asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2}
+)
+
+//////////////////////////////////////
+
+// OidForStandardExtension indicates whether oid identifies a standard extension.
+// Standard extensions are listed in RFC 5280 (and other RFCs).
+func OidForStandardExtension(oid asn1.ObjectIdentifier) bool {
+	if oid.Equal(OidExtensionSubjectKeyId) ||
+		oid.Equal(OidExtensionKeyUsage) ||
+		oid.Equal(OidExtensionExtendedKeyUsage) ||
+		oid.Equal(OidExtensionAuthorityKeyId) ||
+		oid.Equal(OidExtensionBasicConstraints) ||
+		oid.Equal(OidExtensionSubjectAltName) ||
+		oid.Equal(OidExtensionCertificatePolicies) ||
+		oid.Equal(OidExtensionNameConstraints) ||
+		oid.Equal(OidExtensionCRLDistributionPoints) ||
+		oid.Equal(OidExtensionIssuerAltName) ||
+		oid.Equal(OidExtensionSubjectDirectoryAttributes) ||
+		oid.Equal(OidExtensionInhibitAnyPolicy) ||
+		oid.Equal(OidExtensionPolicyConstraints) ||
+		oid.Equal(OidExtensionPolicyMappings) ||
+		oid.Equal(OidExtensionFreshestCRL) ||
+		oid.Equal(OidExtensionSubjectInfoAccess) ||
+		oid.Equal(OidExtensionAuthorityInfoAccess) ||
+		oid.Equal(OidExtensionCTPoison) ||
+		oid.Equal(OidExtensionCTSCT) {
+		return true
+	}
+	return false
+}
+
+// OidInExtensions indicates whether the extension identified by oid is present in extensions.
+func OidInExtensions(oid asn1.ObjectIdentifier, extensions []pkix.Extension) (int, bool) {
+	count := 0
+	critical := false
+	for _, ext := range extensions {
+		if ext.Id.Equal(oid) {
+			count++
+			if ext.Critical {
+				critical = true
+			}
+		}
+	}
+	return count, critical
+}
+
+// String formatting for various X.509/ASN.1 types
+func bitStringToString(b asn1.BitString) string {
+	result := hex.EncodeToString(b.Bytes)
+	bitsLeft := b.BitLength % 8
+	if bitsLeft != 0 {
+		result += " (" + strconv.Itoa(8-bitsLeft) + " unused bits)"
+	}
+	return result
+}
+
+func publicKeyAlgorithmToString(algo x509.PublicKeyAlgorithm) string {
+	switch algo {
+	case x509.RSA:
+		return "rsaEncryption"
+	case x509.DSA:
+		return "dsaEncryption"
+	case x509.ECDSA:
+		return "id-ecPublicKey"
+	default:
+		return strconv.Itoa(int(algo))
+	}
+}
+
+func appendHexData(buf *bytes.Buffer, data []byte, count int, prefix string) {
+	for ii, byte := range data {
+		if ii%count == 0 {
```

Comment:
Could you add a couple of comments within this function to just explain what each part is doing?  I'm sure I could work it out but am currently having to think quite hard about it to do so :P  E.g. what is count here?  When you are checking ii%count ==0, why are you doing that?  etc.